[SPARK-6207] [YARN] [SQL] Adds delegation tokens for metastore to conf.

dougb opened the pull request:
Adds hive2-metastore delegation token to conf when running in secure mode.
Without this change, running on YARN in cluster mode fails with a
GSS exception.

This is a rough patch that adds a dependency to spark/yarn on hive-exec.
I'm looking for suggestions on how to make this patch better.

This contribution is my original work and that I licenses the work to the
Apache Spark project under the project's open source licenses.

Author: Doug Balog doug.balog@target.com

Commit:
[SPARK-6207] [YARN] [SQL] Adds delegation tokens for metastore to conf.
Adds hive2-metastore delagations token to conf when running in securemode.
Without this change, runing on YARN in cluster mode fails with a
GSS exception.

This contribution is my original work and that I licenses the work to the
Apache Spark project under the project's open source licenses.

Author: Doug Balog <doug.balog@target.com>

AmplabJenkins commented:
Can one of the admins verify this patch?

tgravescs commented:
jenkins test this please

tgravescs commented:
thanks for providing the patch.  Which versions of Hive have you tested this with? 

I'm trying to build with a hive 13 version (not the official 0.13.1 though) and see a compile error:
java.lang.AssertionError: assertion failed: org.apache.hadoop.hive.metastore.api.AlreadyExistsException

We will have to find a way to conditionally do the Hive stuff as its optional to compile it in.  I'll have to look at how the hive stuff is done now a bit more to figure out what makes sense.

adding @marmbrus to see if he has ideas as I think his familiar with how the hive stuff is done now.

dougb commented:
Thanks for looking at my patch.
I've tested it only with 0.13.1 from org.spark-project.hive.
I tried to figure out how hive was included via profiles, but gave up.
 @marmbrus, any suggestions would be appreciated.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -903,6 +908,30 @@ object Client extends Logging {
+   * Obtains token for the Hive metastore and adds them to the credentials.
+   */
+  private def obtainTokenForHiveMetastore(conf: Configuration, credentials: Credentials) {
+    if (UserGroupInformation.isSecurityEnabled /* And Hive is enabled */) {
```
So, I think the `/* And Hive is enabled */` part is kinda important, and not talking just about Hive being compiled in. What would happen if a user doesn't have any Hive services running?

I'm also a little worried about precedent here... couldn't we make the same argument for acquiring HBase tokens? Solr? Somthing else? Unfortunately, I can't think of any alternative off the bat, given the way cluster mode works... :-/

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -903,6 +908,30 @@ object Client extends Logging {
+      hive2Token.decodeFromUrlString(tokenStr)
+      credentials.addToken(new Text("hive.server2.delegation.token"), hive2Token)
+      logDebug("Added the Hive Server 2 token to conf.")
+      org.apache.hadoop.hive.ql.metadata.Hive.closeCurrent
```
It looks like most of the existing hive condition compile stuff is just building separate jars and including them. Not ideal here unless we made separate jar for this. Seems like reflection might be the more straight forward way here. 

Also on whether the hive token is required it might make sense to make a config for it (like spark.yarn.hive.token.required).  Similar to how MR has a config to get the job history token if your job is going to talk to the history server.  Its not a nice as far as automatically getting it and we could try other things like checking to see if the hive.metastore.uris config is set but it makes it obvious to the user what is going on and it won't try if not told to.  Thus avoiding issues with people have hive compiled in but not configured or when running app that doesn't require hive access.

One other thing I'm not familiar with and need to look into is the renewal of these tokens.  I know the RM doesn't renew them so is there something in hive that does, if not then the spark app could only run for 24 hours.  Or perhaps they don't expire that often.

The other option is to provide some way for user to fill in the tokens themselves.  Get any tokens they need and have spark add them. As far as I know hive and hbase don't provide command line way to get the tokens though like hdfs does. That would mean you have to write java program to fetch them first.

@vanzin does secure hbase access not work either?  I haven't tried it myself but I'm guessing it doesn't.

tgravescs replied:
so I looked into the hive token renewal. It looks like they last 1 day be default and there is no automatic renewal right now. There is a Hive thrift-call to renew but nothing (like yarn RM) currently automatically calls it.   This means jobs can only run for less then 24 hours.

vanzin replied:
Hbase needs its own delegation token, so I guess it wouldn't work either without some code to fetch the token.

Regarding token expiry, especially for a long running process (let's say the thrift server), it would require something similar to what Hari is doing in #4688 .

dougb replied:
I talked to @marmbrus and @pwendell at the Spark Summit yesterday.
I think I'm going to take a shot at the reflection approach first. 

I would like to see the various delegation tokens collected automatically.
I think it would be a burden for the users to remember to do this in every job.
 
I also don't see anything to renew the namenode tokens. I'm still looking around to see how its handled in other projects.

I think an expert user could add the delegation tokens to the conf by hand,
 if they knew what the config option was, and how to get and encode the token. 

I just started looking deep into spark, but it looks like delegation token management could be better. I need to look at how mapred jobs handle this.

dougb replied:
I'll look at what Hari is doing.

tgravescs replied:
hdfs (namenode) delegation tokens are renewed by the Yarn resourcemanager for you, up til they expire at a week. (Then you need pr4688)  Unfortunately the resourcemanager it doesn't handle hive or hbase tokens.  I personally think putting in this code for hive and then possible hbase for us to know how to get it is ok as long as the interfaces we are using are public and not likely to change.  However we should have a way to skip it if its not configured.

Yes, long running services should be able to renew or reacquire with what Hari is doing.

XuTingjun commented:
I have tested on secure hbase, and it didn't work. On executor process we got the error:

tgravescs commented:
@doubg I'm just curious, are you still working on this?  I think a version of this could go in to atleast allow it to work for the 7 days then Hari's change could improve.

dougb commented:
@tgravescs yes, Sorry, I'll update the pr today. I have a new version using reflection that works in my environment with Hive 0.14.

dougb commented:
@XuTingjun Sorry, this patch is for Hive, not hbase. I'm sure something similar could be created for hbase.

Commit:
Merge branch 'master' of git://github.com/apache/spark into SPARK-6207

Commit:
[SPARK-6207] Second pass at adding Hive delegation token to conf.
- Use reflection instead of adding dependency on hive.
- Tested on Hive 0.13 and Hadoop 2.4.1

tgravescs commented:
ok to test

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -920,6 +925,64 @@ object Client extends Logging {
+        val metastore_uri = hiveConfGet("hive.metastore.uris")
+
+        // Check for local metastore
+        if(metastore_uri != None && metastore_uri.get.toString.size > 0) {
```
add space between if and (

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -920,6 +925,64 @@ object Client extends Logging {
+          val principal = hiveConfGet(metastore_kerberos_principal_conf_var)
+
+          val username = Option(UserGroupInformation.getCurrentUser().getUserName)
+          if(principal != None && username != None){
```
similar add spaces between if and ( and at end between ) and {

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -920,6 +925,64 @@ object Client extends Logging {
+        val hiveConfClass = mirror.classLoader.loadClass("org.apache.hadoop.hive.conf.HiveConf")
+
+        val hiveConfGet = (param:String) => Option(hiveConfClass
+          .getMethod("get",classOf[java.lang.String])
```
add space after ,

tgravescs commented:
A few nits but other then that it looks fine. I"m going to try it out.

Did you test this with hive compiled in and then without it also?

Commit:
[SPARK-6207] Fixes minor code spacing issues.

dougb commented:
Updated code per @tgravescs comments.
I tested with and without hive compiled in and I didn't see any problems.

JoshRosen commented:
Jenkins, this is ok to test.

SparkQA commented:
[Test build #30100 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30100/consoleFull) for   PR 5031 at commit [`3e9ac16`](https://github.com/apache/spark/commit/3e9ac16a8e2e0bdd11f850428115bdc891874b1e).

SparkQA commented:
[Test build #30100 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30100/consoleFull) for   PR 5031 at commit [`3e9ac16`](https://github.com/apache/spark/commit/3e9ac16a8e2e0bdd11f850428115bdc891874b1e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30100/
Test FAILed.

dougb commented:
The Scalastyle checks that failed had nothing to do with this pr.
This pr only changes `yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala`
<pre>
=========================================================================
Running Scala style checks
=========================================================================
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:172:8: Public method must have explicit type
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala:159:15: Public method must have explicit type
[error] (catalyst/compile:scalastyle) errors exist
[error] Total time: 7 s, completed Apr 11, 2015 7:34:10 PM
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:172:8: Public method must have explicit type
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala:159:15: Public method must have explicit type
[error] (catalyst/compile:scalastyle) errors exist
[error] Total time: 6 s, completed Apr 11, 2015 7:34:27 PM
[error] Got a return code of 1 on line 125 of the run-tests script.
Archiving unit tests logs...
</pre>

JoshRosen commented:
I've pushed a hotfix to fix those style errors, so this should hopefully be able to test now.  Let's try again...

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #30101 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30101/consoleFull) for   PR 5031 at commit [`3e9ac16`](https://github.com/apache/spark/commit/3e9ac16a8e2e0bdd11f850428115bdc891874b1e).

SparkQA commented:
[Test build #30101 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30101/consoleFull) for   PR 5031 at commit [`3e9ac16`](https://github.com/apache/spark/commit/3e9ac16a8e2e0bdd11f850428115bdc891874b1e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30101/
Test PASSed.

tgravescs commented:
lgtm.
